src/firebase-social-provider.js

```diff
@@ -0,0 +1,275 @@
+      lastUpdated: Date.now(),
+      url: friend.url || '',
+      imageData: friend.imageData || ''
+    };
+    this.dispatchEvent_(
+        'onUserProfile', this.loginState_.userProfiles[friend.userId]);
+    return;
+  }
+
+  // Ensure that a permanent friend object exists, with the users name.
+  // This must be present in order for other friends to properly detect our
+  // clients, based on the current Firebase rules configuration.
+  var myUrlForFriend =
+      this.baseUrl_ + '/' + this.networkName_ + ':' + this.getUserId_() +
+      '/friends/' + this.networkName_ + ':' + friend.userId;
+  var myRefForFriend = new Firebase(myUrlForFriend);
+  myRefForFriend.set({isFriend: true});
+
+  // Initialize an inbox, writable only by my friend, and unique to this
+  // agent (client).  This should be cleared when I disconnect.
+  var myInboxForFriendRef = new Firebase(
+      myUrlForFriend + '/inbox/' + this.loginState_.agent);
+  myInboxForFriendRef.set('empty');
+  myInboxForFriendRef.onDisconnect().remove();
+
+  // Monitor my new inbox.
+  myInboxForFriendRef.on('child_added', function(value) {
+    value.forEach(function(snapshot) {
+      var fromAgent = snapshot.key();
+      var message = snapshot.val();
+      var clientId = friend.userId + '/' + fromAgent;
+      var fromClientState = this.loginState_.clientStates[clientId];
+      if (!fromClientState) {
+        // We won't have the client yet if the user had never yet signed onto
+        // Firebase until after we logged in.  In this case just add a client
+        // for them.
+        console.log('Got message with unknown client from: ' + friend.userId +
+            ', key: ' + fromAgent + ', val: ' + message);
+        fromClientState =
+            this.addOrUpdateClient_(friend.userId, clientId, 'ONLINE');
+      }
+      fromClientState.lastSeen = Date.now();
+      this.dispatchEvent_(
+          'onMessage', {from: fromClientState, message: message});
+      snapshot.ref().remove();  // Delete message from Firebase server.
+    }.bind(this));
+  }.bind(this));
+
+  // Create a userProfile right away.
+  this.loginState_.userProfiles[friend.userId] = {
+    userId: friend.userId,
+    name: friend.name || '',
+    lastUpdated: Date.now(),
+    url: friend.url || '',
+    imageData: friend.imageData || ''
+  };
+  this.dispatchEvent_(
+      'onUserProfile', this.loginState_.userProfiles[friend.userId]);
+
+  // Get and monitor clients for friend.
+  // Note that if the friend has never yet signed onto Firebase (i.e. their
+  // folder at the directory returned by getClientsUrl_ doesn't exist yet)
+  // this will never return any clients.  In that case we will only know
+  // about a client when they send us the first message (e.g. instance message
+  // in the case of uProxy).
+  var clients = new Firebase(this.getClientsUrl_(friend.userId));
+  clients.on('value', function(value) {
+    // TODO: this is going to be updating the lastUpdated and lastSeen values
```

Comment:
Strongly agreed with this TODO - I'm okay with not resolving it prior to merging this, but it may be worth adding the url `See: https://www.firebase.com/docs/web/api/query/on.html` to this comment as a place to start for whoever looks at this.